Server/src/main/java/org/gluu/oxauth/token/ws/rs/TokenRestWebServiceImpl.java

```diff
@@ -490,6 +495,103 @@ public Response requestAccessToken(String grantType, String code,
+
+        log.trace("DeviceCodeGrant : '{}'", deviceCodeGrant);
+
+        if (deviceCodeGrant != null) {
+            if (!deviceCodeGrant.getClientId().equals(client.getClientId())) {
+                return response(error(400, TokenErrorResponseType.INVALID_GRANT, "The client is not authorized."), oAuth2AuditLog);
+            }
+            if (!deviceCodeGrant.isTokensDelivered()) {
+                RefreshToken refToken = deviceCodeGrant.createRefreshToken();
+                log.debug("Issuing refresh token: {}", refToken.getCode());
+
+                AccessToken accessToken = deviceCodeGrant.createAccessToken(request.getHeader("X-ClientCert"), new ExecutionContext(request, response));
+                log.debug("Issuing access token: {}", accessToken.getCode());
+
+                IdToken idToken = deviceCodeGrant.createIdToken(
+                        null, null, accessToken, refToken,
+                        null, deviceCodeGrant, false, null);
+
+                deviceCodeGrant.setTokensDelivered(true);
+                deviceCodeGrant.save();
+
+                RefreshToken reToken = null;
+                if (isRefreshTokenAllowed(client, deviceCodeGrant)) {
+                    reToken = refToken;
+                }
+
+                if (scope != null && !scope.isEmpty()) {
+                    scope = deviceCodeGrant.checkScopesPolicy(scope);
+                }
+                log.info("Device authorization in token endpoint processed and return to the client, device_code: {}", deviceCodeGrant.getDeviceCode());
+
+                oAuth2AuditLog.updateOAuth2AuditLog(deviceCodeGrant, true);
+
+                return Response.ok().entity(getJSonResponse(accessToken, accessToken.getTokenType(),
+                        accessToken.getExpiresIn(), reToken, scope, idToken)).build();
+            } else {
+                return response(error(400, TokenErrorResponseType.INVALID_GRANT, "AuthReqId is no longer available."), oAuth2AuditLog);
+            }
+        } else {
+            final DeviceAuthorizationCacheControl cacheData = deviceAuthorizationService.getDeviceAuthorizationCacheData(deviceCode, null);
+            log.trace("DeviceAuthorizationCacheControl data : '{}'", cacheData);
+            if (cacheData != null) {
+                if (!cacheData.getClient().getClientId().equals(client.getClientId())) {
+                    return response(error(400, TokenErrorResponseType.INVALID_GRANT, "The client is not authorized."), oAuth2AuditLog);
+                }
+                long currentTime = new Date().getTime();
+                Long lastAccess = cacheData.getLastAccessControl();
+                if (lastAccess == null) {
+                    lastAccess = currentTime;
+                }
+                cacheData.setLastAccessControl(currentTime);
+                deviceAuthorizationService.saveInCache(cacheData, true, true);
+
+                if (cacheData.getStatus() == DeviceAuthorizationStatus.PENDING) {
```

Comment:
Minor : All validation checks can be written in form of WebApplicationException and be linear instead of if/else with another inner if/else. It's not always possible to get it linear but here it seems to be possible.